Added Forge Multipart

amadornes opened the pull request:
It's been in the works for over half a year but it's finally here! Forge Multipart (formerly known as MCMultiPart) is being PRed to Forge! :D
I @amadornes consent to the licensing of my contributions to the Forge project under the license as stated in LICENSE-new.txt

border999 commented:
This is the ship of all forge modders. Just shipping forge and Forge Multipart.

Actuarius commented:
@williewillus added labels [Feature]

Rumaruka commented:
add Lua :D

ProfessorProspector commented:
http://i.imgur.com/LRMovuR.gif

TheCodedOne commented:
About time! :smiley:

loordgek commented:
need more gifs http://mashable.com/2013/07/09/excitement-gifs/#6YGCPypH7aq5

way2muchnoise commented:
Waiting for that moment it gets merged
![](http://puu.sh/p1au3/27c75d5342.gif)

tterrag1098 commented:
As potentially exciting as this is...it's quite a bit of code with no explanations. Do you have some API docs or usage examples?

VelocityRa commented:
Someone requested more gifs I believe
![giphy](https://cloud.githubusercontent.com/assets/6632271/15455705/14f72192-2065-11e6-8586-d4ca3b42fce6.gif)

nabijaczleweli commented:
:shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit::shipit:

shadowfacts commented:
@tterrag1098 There are a number of open source mods using it (not this specific version, obviously):

- [CraftingSlabs](https://github.com/shadowfacts/craftingslabs)
- [Chisels & Bits](https://github.com/AlgorithmX2/Chisels-and-Bits)
- [Simple Covers](https://github.com/bdew/covers)
- [Pressure Pipes](https://github.com/bdew/pressure/tree/mc1.9.0)

And probably more that I don't know of

williewillus commented:
Yeah I'm also hyped but don't want another system with barely any docs (coughfry ;P) 
If you could make some readthedocs articles for a quick start, covering core concepts etc., that would be fantastic

UnRealDinnerbone commented:
Also [Charset](github.com/asiekierka/CharsetMC)

topisani commented:
to add to that list:
 - [Charset] (https://github.com/asiekierka/CharsetMC)
 - [Neotech] (https://github.com/TeamBR-Modding/NeoTech)
 - [TechReborn] (https://github.com/TechReborn/TechReborn)

DiMuRie commented:
yay multiparts

VelocityRa commented:
@t1g3l  There are default implementations for all interfaces, and examples on how to register them are [here](https://github.com/amadornes/MCMultiPart/blob/1.9/src/main/java/mcmultipart/MCMultiPartMod.java#L32)

asiekierka commented:
@williewillus amadornes usually recommends Charset's source code as a good set of examples on how the API works

DiMuRie commented:
@asiekierka yup

RainWarrior commented on src/main/java/net/minecraftforge/fmp/ForgeMultipart.java:
```diff
@@ -0,0 +1,93 @@
+import net.minecraftforge.fmp.network.MultipartNetworkHandler;
+import net.minecraftforge.fmp.util.MCMPEventHandler;
+
+@Mod(modid = ForgeMultipart.MODID, name = ForgeMultipart.NAME, version = ForgeMultipart.VERSION)
```
Usually things like these hooks go into FML sided handlers directly. Lex/cpw should know more.

amadornes replied:
Already talked to @cpw. Working on it!

GotoLink commented on src/main/java/net/minecraftforge/fmp/ForgeMultipart.java:
```diff
@@ -0,0 +1,93 @@
+        GameRegistry.register(multipart, new ResourceLocation(ForgeMultipart.MODID, "multipart"));
+        GameRegistry.registerTileEntityWithAlternatives(TileMultipartContainer.Ticking.class, MODID + ":multipart.ticking",
+                "mcmultipart:multipart.ticking", "mcmultipart:multipart");
+        GameRegistry.registerTileEntityWithAlternatives(TileMultipartContainer.class, MODID + ":multipart.ticking",
```
Conflict in name here. Should probably be 'nonticking'.

amadornes replied:
Oh, whoops... That was nonticking indeed!

GotoLink commented on src/main/java/net/minecraftforge/fmp/block/BlockMultipartContainer.java:
```diff
@@ -0,0 +1,561 @@
+
+    @SubscribeEvent(priority = EventPriority.HIGH)
+    @SideOnly(Side.CLIENT)
+    public final void onDrawBlockHighlight(DrawBlockHighlightEvent event)
```
Is there a reason for this handler to be in every block instance that could be created from here ?

asiekierka replied:
Oof, that is ugly. I see no reason - should be factored out.

amadornes replied:
Will do!

GotoLink commented on src/main/java/net/minecraftforge/fmp/network/MessageMultipartChange.java:
```diff
@@ -0,0 +1,187 @@
+    @SideOnly(Side.CLIENT)
+    private void schedulePacketHandling(final MessageMultipartChange message)
+    {
+        Minecraft.getMinecraft().addScheduledTask(new Runnable()
```
I would use FMLCommonHandler#getWorldThread(INetHandler) instead. Avoid hard link with Minecraft client class.

amadornes replied:
I'll look into it :)

GotoLink commented on src/main/java/net/minecraftforge/fmp/block/BlockMultipartContainer.java:
```diff
@@ -0,0 +1,561 @@
+    public boolean addLandingEffects(IBlockState state, WorldServer worldObj, BlockPos blockPosition, IBlockState iblockstate,
+            EntityLivingBase entity, int numberOfParticles)
+    {
+        return true;
```
Why not pass to the parts ? And why cancel the landing effect by default ?

amadornes replied:
I tried to figure out which part the player was standing on but it turned out to not be so easy, so I just canceled the event so the default stone particles wouldn't spawn.

LexManos replied:
May be something to look into, but it could be just as simple as a raytrace of the player's center position straight down. But that's not important right now, just something to keep in mind.

asiekierka replied:
Not really - one has to keep in mind that the player could be standing near the edge of a hole, and we don't want to raytrace into the hole.

amaranth replied:
They landed on this block so you know you don't need to trace further than that.

tterrag1098 commented:
Examples are nice but I agree with @williewillus, explanation docs for at least the basics would be good.

amadornes commented:
@tterrag1098 Docs in a wiki will be a thing very soon. The javadocs in the classes are quite helpful, though, so at least you have that for now :)

LexManos commented on patches/minecraft/net/minecraft/util/math/RayTraceResult.java.patch:
```diff
@@ -13,3 +13,29 @@
++        ENTITY,
++        MULTIPART{ public boolean isBlock() { return true; } };
++        public boolean isBlock() { return false; }
+     }
```
Absolutely none of these patches are necessary.

asiekierka replied:
They are, however, very convienent to have - otherwise, we get into the joys and wonders of typecasting hell.

amadornes replied:
As asie said, even though they aren't necessary, it's much more convenient doing multipart raytracing using the normal raytrace result than having to typecast all over the place to make sure you're looking at a multipart. I've talked to @cpw about it and he's okay with the change. You can look at the original code for yourself and see how messy it was :P

GotoLink replied:
I'd like to suggest reducing patch size with
`public boolean isBlock(){return this == BLOCK || this == MULTIPART;}`

LexManos commented on src/main/java/net/minecraftforge/fmp/ForgeMultipart.java:
```diff
@@ -0,0 +1,58 @@
+
+        proxy.preInit();
+
+        MinecraftForge.EVENT_BUS.register(new MCMPEventHandler());
```
Why is this a seperate mod, if the API is in place make it like the Universal bucket, wait until a mod opt's in to the system and then have Forge itself register all the necessary parts. Under forge's name.

amadornes replied:
Talked to @cpw about it. It is now an FML mod, just like FML, Forge and MCP. Version is the same as Forge's because I don't need anything fancy for it. We should be able to merge it into Forge's mod class, but it'd only make it more confusing and, atm, I'm trying to keep FMP as a separate "entity" in the system as possible, in case anything happens and it needs to be modified.

LexManos commented on src/main/java/net/minecraftforge/fmp/block/BlockCoverable.java:
```diff
@@ -0,0 +1,730 @@
+ * player isn't interacting with a multipart.<br/>
+ * Extend {@link TileCoverable} or implement {@link IMicroblockContainerTile} and return it in
+ * {@link BlockCoverable#createNewTileEntity(World, int)} if you want a custom tile entity for your block.
+ */
```
This all seems like a sidestep of our capability system... We don't want people implementing interfaces anymore in the main TE/Block classes unless there is a MAJOR reason to do it.

asiekierka replied:
MAJOR reason to do it - BlockCoverable hooks almost every method in Block.

amadornes replied:
The purpose of BlockCoverable is to allow people to keep using a normal block but have the ability to place microblocks in the same space as that block. It's a "light" solution to adding multipart support that, while keeping your Block and TE like they were, allows you to have the extra microblock support most mods want.
The only way I can think of that would be able to get around this (without removing it completely) would be some kind of "layered logic" system where the default behavior (vanilla's) would be the bottommost layer and developers would be able to build logic on top of that which falls back to the next layer if the current one doesn't support it. This way, there would be the "vanilla" layer, then the "microblock container" layer and finally all the blocks' logic layers. I'm not sure how reliable and fast a system like that would be, but I think this is a fairly simple and good solution for now.

LexManos commented on src/main/java/net/minecraftforge/fmp/block/BlockCoverable.java:
```diff
@@ -0,0 +1,730 @@
+    }
+
+    protected IMicroblockContainerTile getMicroblockTile(IBlockAccess world, BlockPos pos)
+    {
```
Instanceof nightmare?

amadornes replied:
Black magic using generics. It forced you to return an object that extended TileEntity and implemented IMicroblockContainerTile. I'll get rid of it in favour of capabilities.

LexManos commented on src/main/java/net/minecraftforge/fmp/block/BlockCoverable.java:
```diff
@@ -0,0 +1,730 @@
+            return hit;
+        }
+        bounds = result.boxHit;
+        return result;
```
instance variable in a block... eww...

amadornes replied:
This was the only way to handle something like this, tbh. When the selection box is retrieved, no external information is provided, so you have to save it when you raytrace it and provide it later. Not the cleanest solution, but I'd blame vanilla for it.

LexManos commented on src/main/java/net/minecraftforge/fmp/block/BlockCoverable.java:
```diff
@@ -0,0 +1,730 @@
+    @Override
+    public final ItemStack getPickBlock(IBlockState state, RayTraceResult target, World world, BlockPos pos, EntityPlayer player)
+    {
+        if (target instanceof RayTraceResult)
```
Wat......

amadornes replied:
Hah, I missed that one when converting from PartMOP to the modified version of RayTraceResult, but if that's not a valid change I guess I'll have to revert it back, which is going to be fun...

LexManos commented on src/main/java/net/minecraftforge/fmp/block/BlockCoverable.java:
```diff
@@ -0,0 +1,730 @@
+    {
+        if (harvestingWrapper) { return getDropsDefault(world, pos, state, fortune); }
+        IMicroblockContainerTile brokenTile = getMicroblockTile(world, pos);
+        if (brokenTile == null)
```
Don't shade names like this on purpose... Also instance variables on Block is not a good idea..... Threads kill it...

LexManos commented on src/main/java/net/minecraftforge/fmp/block/BlockCoverable.java:
```diff
@@ -0,0 +1,730 @@
+    {
+        RayTraceResult hit = reTraceAll(world, pos, player);
+        if (hit instanceof RayTraceResult)
+        {
```
Again?

amadornes replied:
Yeah... leftovers from PartMOP... You can see I had to do a ton of typecasting all over the place...

LexManos commented on src/main/java/net/minecraftforge/fmp/block/BlockCoverable.java:
```diff
@@ -0,0 +1,730 @@
+            }
+            else
+            {
+                if (!removedByPlayerDefault(state, world, pos, player, willHarvest)) { return false; }
```
better to put on new line then to wrap in {}s and stick on the end.

amadornes replied:
Will do! I just ran all the code through a formatter that followed the guidelines cpw gave me and that was an option I wasn't quite sure how you wanted to handle, so I just did brackets because you like them. Them appearing on the same line... was more of a dep. If you don't mind one line stuff after ifs and fors without brackets, I'm more than happy to do that, since that's what I do. But I'd like you to confirm it first, I don't want to mess everything up... again...

LexManos commented on src/main/java/net/minecraftforge/fmp/block/BlockCoverable.java:
```diff
@@ -0,0 +1,730 @@
+                world.removeTileEntity(pos);
+                if (!world.setBlockState(pos, ForgeMultipart.multipart.getDefaultState(), 0)) { return false; }
+                world.removeTileEntity(pos);
+                world.setTileEntity(pos, new TileMultipartContainer(container));
```
Why are you removing twice, just to add again.... what?

AtomicBlom replied:
I agree @amadornes, using nested if statements would make this code a lot more readable.

amadornes replied:
Removing it twice to add it again, yes... but no. The TE as it is created by default has a new multipart container. What I want to do is clone the old one onto the one that's just been created. I'll use a copyFrom() method instead, I guess. May be a bit better for memory, though here I was just reusing the instance.

LexManos commented on src/main/java/net/minecraftforge/fmp/block/BlockCoverable.java:
```diff
@@ -0,0 +1,730 @@
+
+    @Override
+    public final Boolean isAABBInsideMaterial(World world, BlockPos pos, AxisAlignedBB aabb, Material material)
+    {
```
Out of curiosity as I don't have the code in front of me, what does null signify and why do we care to have this a ternary return?

amadornes replied:
cc: @AlgorithmX2

AlexIIL replied:
Copied directly from the javadoc in ````Block````
````java
 /**
      * Called when boats or fishing hooks are inside the block to check if they are inside
      * the material requested.
      *
      * @param world world that is being tested.
      * @param pos block thats being tested.
      * @param boundingBox box to test, generally the bounds of an entity that are besting tested.
      * @param materialIn to check for.
      * @return null for default behavior, true if the box is within the material, false if it was not.
      */
````

LexManos commented on src/main/java/net/minecraftforge/fmp/block/BlockCoverable.java:
```diff
@@ -0,0 +1,730 @@
+        IMicroblockContainerTile tile = getMicroblockTile(world, pos);
+        MicroblockContainer container = tile != null ? tile.getMicroblockContainer() : null;
+        if (container == null) { return false; }
+        return container.getPartContainer().canPlaceTorchOnTop() || canPlaceTorchOnTopDefault(state, world, pos);
```
All of these early returns you do like this need to be switched to if (!=null)'s 1) it's cleaner and 2) it makes code easier to understand and 3) it makes future expansion easier without need to refactor these two lines.

Not to mention all you single line ifs need to be converted into two lined one minus the braces.

LexManos commented on src/main/java/net/minecraftforge/fmp/block/BlockCoverable.java:
```diff
@@ -0,0 +1,730 @@
+        randomDisplayTickDefault(state, world, pos, rand);
+        IMicroblockContainerTile tile = getMicroblockTile(world, pos);
+        MicroblockContainer container = tile != null ? tile.getMicroblockContainer() : null;
+        if (container != null)
```
Again this would all be cleaner if you didnt have that intermediate IMicroblockContainerTile and just used caps.

asiekierka replied:
That intermediate IMicroblockContainerTile would have to be there anyway, as the expectation is that the TileEntity overrides a lot of methods to let those microblocks exist and behave correctly.

amadornes replied:
While asie is right in saying that you are supposed to implement your own implementations for the methods in IMicroblockContainerTile, I agree that it should be turned into a capability. It will not have a default implementation, though, unless you want me to be hacky and check microblock placement against the original block's collision boxes.

LexManos commented on src/main/java/net/minecraftforge/fmp/block/TileCoverable.java:
```diff
@@ -0,0 +1,169 @@
+    @Override
+    public boolean hasCapability(Capability<?> capability, EnumFacing facing)
+    {
+        if (super.hasCapability(capability, facing)) { return true; }
```
Typically this is the other way around. The specific handler, and then the super, because the super contains the handlers added by the events from other mods.

amadornes replied:
Hmm... Okay, should be very easy to fix

LexManos commented on src/main/java/net/minecraftforge/fmp/block/TileCoverable.java:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    @Override
+    public <T> T getCapability(Capability<T> capability, PartSlot slot, EnumFacing facing)
```
What does this give that the normal cap system doesn't?

asiekierka replied:
Access to more precise locations in the multipart. Let's say you want a capability from the corner of a block to another corner.

amadornes replied:
Normal capabilities are only provided based on the side that they're being requested for. What this allows you to do is have one part that occupies the top of the block and one that occupies the bottom, for example, and have each one return a different set of capabilities for all its sides. This way external objects can interact with the capabilities on lets say an edge, without having to get the specific part at that position, make sure it's not null and then get the capability.

LexManos commented on src/main/java/net/minecraftforge/fmp/block/TileMultipartContainer.java:
```diff
@@ -0,0 +1,352 @@
+    }
+
+    @Override
+    public Collection<? extends IMultipart> getParts()
```
All these bouncer methods give me a migraine -.-

amadornes replied:
Yes, they'll go away when I turn IMultipartContainer into a cap :)

LexManos commented on src/main/java/net/minecraftforge/fmp/block/TileMultipartContainer.java:
```diff
@@ -0,0 +1,352 @@
+
+            for (IMultipart part : getParts())
+            {
+                if (part instanceof ITickable)
```
Better to have this baked instead of doing it each tick...

amadornes replied:
Will do :)

LexManos commented on src/main/java/net/minecraftforge/fmp/capabilities/CapabilityWrapperItemHandler.java:
```diff
@@ -0,0 +1,124 @@
+import net.minecraftforge.items.IItemHandlerModifiable;
+
+public class CapabilityWrapperItemHandler implements ICapabilityWrapper<IItemHandler>
+{
```
We already have these classes, https://github.com/MinecraftForge/MinecraftForge/tree/master/src/main/java/net/minecraftforge/items/wrapper

asiekierka replied:
We don't. This joins together all the IItemHandlers accessible from a given side of the multipart.

amadornes replied:
This is meant to take multiple implementations of the same capability and wrap them into just one object which can be returned when that cap is requested on the side of the multipart TE. This way other developers that just want to access the parts' inventory, for example, can insert items normally without having to adapt their code to iterate through all of them.

LexManos commented on src/main/java/net/minecraftforge/fmp/capabilities/CapabilityWrapperRegistry.java:
```diff
@@ -0,0 +1,34 @@
+
+import net.minecraftforge.common.capabilities.Capability;
+
+public class CapabilityWrapperRegistry
```
Why? Doing this is BAD as it will cause SEVERE lag when querying....

amadornes replied:
Where exactly? The comment is on line 9 and that's just the class declaration lol.
I get what you mean, though. Wrapping all the caps when they are requested... But that's how caps work, you don't know if they have changed until you check.
What else do you want me to do? Put all the implementations into a weak set alongside the wrapped version and whenever it's requested, check if the amount of implementations and their hashcodes match, if so return the cached one, if not gen a new one and cache it? It's... doable... but it'd be messy af.

LexManos commented on src/main/java/net/minecraftforge/fmp/capabilities/ICapabilityWrapper.java:
```diff
@@ -0,0 +1,14 @@
+
+import net.minecraftforge.common.capabilities.Capability;
+
+public interface ICapabilityWrapper<T>
```
No explanation or javadocs... no....

LexManos commented on src/main/java/net/minecraftforge/fmp/capabilities/ISlottedCapabilityProvider.java:
```diff
@@ -0,0 +1,14 @@
+import net.minecraftforge.common.capabilities.Capability;
+import net.minecraftforge.fmp.multipart.PartSlot;
+
+public interface ISlottedCapabilityProvider
```
No explanation or javadocs... no....

LexManos commented on src/main/java/net/minecraftforge/fmp/capabilities/MultipartCapabilityHelper.java:
```diff
@@ -0,0 +1,175 @@
+/**
+ * A general use multipart capability helper.
+ */
+public class MultipartCapabilityHelper
```
This whole class needs explination, as it seems like an abomination of the capability system for no particular reason...

asiekierka replied:
Multiparts do not just have sides. They also have faces. The class allows more specific access for that.

amadornes replied:
What asie said. This is here to expand upon the (relatively limited) default cap system, as blocks may want to access the caps on only one edge of the block and not the whole TE. It's also much nicer than testing each one manually.

LexManos commented on src/main/java/net/minecraftforge/fmp/capabilities/PartAttachCapabilitiesEvent.java:
```diff
@@ -0,0 +1,22 @@
+import net.minecraftforge.fmp.multipart.IMultipart;
+
+public class PartAttachCapabilitiesEvent extends AttachCapabilitiesEvent
+{
```
If this is needed it'd be moved into the subclass of the event just like everything else not it's own class.

amadornes replied:
Already done! Completely forgot about it

LexManos commented on src/main/java/net/minecraftforge/fmp/client/FMPClientProxy.java:
```diff
@@ -0,0 +1,64 @@
+                new ModelResourceLocation(Block.REGISTRY.getNameForObject(ForgeMultipart.multipart), "ticking=false"),
+                new ModelMultipartContainer(null, null));
+        event.getModelRegistry().putObject(
+                new ModelResourceLocation(Block.REGISTRY.getNameForObject(ForgeMultipart.multipart), "ticking=true"),
```
ForgeMultipart.multipart.getRegistryName()

amadornes replied:
Yup... I discovered that like half an hour ago. But thanks :P

LexManos commented on src/main/java/net/minecraftforge/fmp/client/microblock/IMicroModelProvider.java:
```diff
@@ -0,0 +1,21 @@
+ * Interface that allows you to provide fully custom quads for a specific material.
+ */
+public interface IMicroModelProvider
+{
```
Pinging @RainWarrior for this.

LexManos commented on src/main/java/net/minecraftforge/fmp/client/microblock/MicroblockRegistryClient.java:
```diff
@@ -0,0 +1,30 @@
+{
+    
+    private static final Map<IMicroMaterial, IMicroModelProvider> materialModelProviders = new HashMap<IMicroMaterial, IMicroModelProvider>();
+
```
Pinging @RainWarrior for this.

LexManos commented on src/main/java/net/minecraftforge/fmp/client/multipart/AdvancedParticleManager.java:
```diff
@@ -0,0 +1,233 @@
+import net.minecraftforge.fml.relauncher.SideOnly;
+
+@SideOnly(Side.CLIENT)
+public class AdvancedParticleManager extends ParticleManager
```
Why exactly is this class needed for anything?

asiekierka replied:
To allow creating particle effects with a TextureAtlasSprite, as multipart models are not in the block model map.

amadornes replied:
It's mainly to support custom-textured breaking particles without having to AT the particle constructor.

LexManos commented on src/main/java/net/minecraftforge/fmp/client/multipart/FastMSR.java:
```diff
@@ -0,0 +1,52 @@
+ * Multipart equivalent of {@link FastTESR}.
+ */
+public abstract class FastMSR<T extends IMultipart> extends MultipartSpecialRenderer<T>
+{
```
Pinging @RainWarrior for this. 
The bulk of the rendering should be done through he model system what needs a TESR/'Fast' TESR?

asiekierka replied:
Items inside pipes, for example?

amadornes replied:
I know we're trying to phase out TESRs, but things that need any kind of animation require them atm, as well as things that use shaders or custom OpenGL blending.

asiekierka replied:
Also, vanilla still has four items which need TESRs to render, Also, isn't the animation API effectively a FastTESR?

LexManos commented on src/main/java/net/minecraftforge/fmp/client/multipart/ICustomHighlightPart.java:
```diff
@@ -0,0 +1,27 @@
+     * Draws this part's highlight.
+     */
+    @SideOnly(Side.CLIENT)
+    public boolean drawHighlight(RayTraceResult hit, EntityPlayer player, float partialTicks);
```
Why not just have this part of IMultipart?

asiekierka replied:
Because MCMultiPart provides a default implementation.

amadornes replied:
I have nothing against it, but I wanted to keep as many clientside-only methods outside of IMultipart. I'll throw it in, though.

LexManos commented on src/main/java/net/minecraftforge/fmp/client/multipart/IFastMSRPart.java:
```diff
@@ -0,0 +1,17 @@
+     * Checks whether or not this part currently needs a {@link FastMSR} or it should just use a
+     * {@link MultipartSpecialRenderer}.
+     */
+    public boolean hasFastRenderer();
```
Why not just have this part of IMultipart?

asiekierka replied:
Good idea - this was probably done as the interface was introduced mid-1.9 and might've broken mods?

amadornes replied:
This was indeed left over from the 1.9 version when I finally figured out how to do FastTESRs for multiparts. I'll throw it into IMultipart.

LexManos commented on src/main/java/net/minecraftforge/fmp/client/multipart/MultipartRegistryClient.java:
```diff
@@ -0,0 +1,63 @@
+public class MultipartRegistryClient
+{
+    
+    private static Map<ResourceLocation, IStateMapper> specialMappers = new HashMap<ResourceLocation, IStateMapper>();
```
NOOOOOOO
Having these maps in the rendering pipeline is a HUGE waste.. there has to be a better way.

amadornes replied:
The state mappers are unavoidable and they don't cause too big of an impact because they are only ran when the resources are reloaded. The MSRs, though... just use the same system as TESRs. I could make them use a string as an ID, but the difference won't be huge. If you have a better alternative, I'm open to suggestions :)

LexManos commented on src/main/java/net/minecraftforge/fmp/event/PartEvent.java:
```diff
@@ -0,0 +1,36 @@
+import net.minecraftforge.fmp.multipart.IMultipart;
+
+public abstract class PartEvent extends Event
+{
```
Needs docs, also getters instead of public fields.

amadornes replied:
Done!

LexManos commented on src/main/java/net/minecraftforge/fmp/item/IItemMultipartFactory.java:
```diff
@@ -0,0 +1,16 @@
+import net.minecraftforge.fmp.multipart.IMultipart;
+
+public interface IItemMultipartFactory
+{
```
Docs

LexManos commented on src/main/java/net/minecraftforge/fmp/item/IItemSaw.java:
```diff
@@ -0,0 +1,10 @@
+import net.minecraft.item.ItemStack;
+
+public interface IItemSaw
+{
```
Docs, does this even need to be part of this PR? Or, should this bee moved to a capability...

asiekierka replied:
Seems pointless to me, perhaps a leftover from amadornes's other mod...

amadornes replied:
It's not a leftover, no. It should be moved to a capability, but it's basically an interface that represents anything that can cut a block into microblocks so that anybody can add their own saw and they are all compatible (the microblocks themselves aren't included in the PR for obvious reasons, but in the case that there are two or more microblock mods, this would also be good for inter-compatibility).

asiekierka replied:
Why not just use a ToolClass in the Item instead of an interface?

LexManos commented on src/main/java/net/minecraftforge/fmp/item/ItemMultipart.java:
```diff
@@ -0,0 +1,83 @@
+ * override {@link ItemMultipart#consumeItem(ItemStack)} and {@link ItemMultipart#getPlacementSound(ItemStack)} to place
+ * your part in the world.
+ */
+public abstract class ItemMultipart extends Item implements IItemMultipartFactory
```
How is this used? Do modders subclass this for all of their smaller parts?

asiekierka replied:
It is essentially a default implementation of an item which places a multipart, much like ItemBlock is a default implementation of an item which places a block - however, it is not necessary and one could place the multipart themselves in onItemUse.

amadornes replied:
That is right, yes. Modders extend it and they have a method which gives them all the information they could possibly need to place their multipart. I was considering the idea of making another version of this class (one that extended it) which would just take a factory as an argument for J8 lambda support, but I'd like feedback on that idea.

LexManos commented on src/main/java/net/minecraftforge/fmp/item/PartPlacementWrapper.java:
```diff
@@ -0,0 +1,212 @@
+import net.minecraftforge.fmp.network.MessageWrappedPartPlacement;
+import net.minecraftforge.fmp.raytrace.RayTraceUtils;
+
+public class PartPlacementWrapper
```
Why is this wrapper needed... what does it do?

amadornes replied:
It handles the placement of things like multiparts in the case that they are using an ItemBlock, for example. So if by default it's a block (which can be converted to a part and back), if you attempt to place it inside a block that already has multiparts it'll place a part instead. There's also an adapted version for microblock containers so that they can be placed if there are already microblocks in that blockspace.

LexManos commented on src/main/java/net/minecraftforge/fmp/microblock/BlockMicroMaterial.java:
```diff
@@ -0,0 +1,179 @@
+ * @see IMicroMaterial
+ * @see MicroblockRegistry
+ */
+public class BlockMicroMaterial implements IMicroMaterial
```
What is this for?

amadornes replied:
It's an IMicroMaterial which inherits the properties of the IBlockState that you pass in. Basically a simple block -> microblock material wrapper.

LexManos commented on src/main/java/net/minecraftforge/fmp/microblock/IMicroblockContainerTile.java:
```diff
@@ -0,0 +1,34 @@
+ * this to your {@link TileEntity} using {@link Capability Capabilities}.
+ *
+ * @see IMicroblock
+ * @see IMultipartContainer
```
You say this but the rest of your code base uses this as a instanceof check.. and doesn't register this as a capability..

amadornes replied:
Yeeeeeeeeeahhhh... I may or may not have forgotten... It's already fixed, though :)

LexManos commented on src/main/java/net/minecraftforge/fmp/microblock/IMicroblockPlacementGrid.java:
```diff
@@ -0,0 +1,22 @@
+ * @see IMicroblock
+ * @see MicroblockClass
+ */
+public interface IMicroblockPlacementGrid
```
Why not just sideonly the class? It's a client side, where is this used need more information.

amadornes replied:
I'll just sideonly it, yes. That was temporary because I was considering adding some extra methods which didn't get added in the end. It's not called anywhere, it's just there to represent how a microblock class' placement grid should be rendered. The idea is that any mod will be able to come along and register their own microblock classes (like covers, fences or walls) and whichever mod adds the microblocks will automatically take care of that and register an item to handle it.

LexManos commented on src/main/java/net/minecraftforge/fmp/block/BlockMultipartContainer.java:
```diff
@@ -0,0 +1,511 @@
+    @Override
+    protected BlockStateContainer createBlockState()
+    {
+        return new ExtendedBlockState(this, PROPERTIES, UNLISTED_PROPERTIES);
```
Eww use the builder

amadornes replied:
Builder? I had no clue there was one...

LexManos commented on src/main/java/net/minecraftforge/fmp/microblock/Microblock.java:
```diff
@@ -0,0 +1,541 @@
+        PROPERTIES[0] = PROPERTY_MATERIAL = new PropertyMicroMaterial("material");
+        PROPERTIES[1] = PROPERTY_MATERIAL_STATE = new PropertyBlockState("material_state");
+        PROPERTIES[2] = PROPERTY_SIZE = new PropertyAdapter<Integer>(PropertyInteger.create("size", 0, 7));
+        PROPERTIES[3] = PROPERTY_SLOT = new PropertyAdapter<PartSlot>(PropertyEnum.create("slot", PartSlot.class));
```
eww, use the builder...

LexManos commented on src/main/java/net/minecraftforge/fmp/microblock/MicroblockContainer.java:
```diff
@@ -0,0 +1,135 @@
+    @SuppressWarnings("unchecked")
+    @Override
+    public Collection<? extends IMicroblock> getParts()
+    {
```
Where does it get that IMucroblock from? I see no generics defined anywhere.

tterrag1098 replied:
There's no generic type here, only a wildcard

amadornes replied:
Since MicroblockContainer implements IMultipartContainer and that one returns Collection<? extends IMultipart> I can just do Collection<? extends IMicroblock> so people can iterate through them directly knowing that they are microblocks, without having to cast every single one manually.

LexManos commented on src/main/java/net/minecraftforge/fmp/microblock/MicroblockDelegate.java:
```diff
@@ -0,0 +1,174 @@
+    /**
+     * Called when a player right-clicks this microblock. Return true to play the right-click animation.
+     */
+    public Optional<Boolean> onActivated(EntityPlayer player, EnumHand hand, ItemStack heldItem, RayTraceResult hit)
```
false == what
missing == ?

tterrag1098 replied:
And why not use the now superior ActionResult?

amadornes replied:
true = animate arm
false = click failed
null = pass through to the microblock itself, then to whatever is needed

When this was added EnumActionResult wasn't a thing yet, but I'll switch over to it, yes :)

LexManos commented on src/main/java/net/minecraftforge/fmp/microblock/MicroblockDelegate.java:
```diff
@@ -0,0 +1,174 @@
+ * Represents an object to which microblock events are delegated. This allows for custom interaction and data storage
+ * inside the microblock.
+ */
+public class MicroblockDelegate
```
Why is this class needed? All it does is null everything except sendUpdatePacket?

amadornes replied:
As the javadoc says, if you extend this class and return it in your IMicroMaterial, you'll be able to catch microblock events such as right clicks. There's a basic implementation for crafting tables in the material registry.

LexManos commented on src/main/java/net/minecraftforge/fmp/multipart/Multipart.java:
```diff
@@ -0,0 +1,520 @@
+        BlockPos pos = getPos();
+        double x = pos.getX() + 0.5, y = pos.getY() + 0.5, z = pos.getZ() + 0.5;
+
+        if ((player == null || !player.capabilities.isCreativeMode) && !world.isRemote && world.getGameRules().getBoolean("doTileDrops")
```
.. this should go through getDrops and now spawn the entities directly should it not?

amadornes replied:
No, because on a multipart block getDrops() is never called when harvested by a player - only if something like a block breaker does it (and if that's the case, it'll get all the parts' drops and spawn them at the same time).

LexManos commented on src/main/java/net/minecraftforge/fmp/multipart/MultipartContainer.java:
```diff
@@ -0,0 +1,751 @@
+ * Helper class that contains all the logic required for an {@link IMultipartContainer} to work, as well as methods that
+ * are forwarded to each of the parts.<br/>
+ * You can implement {@link IMultipartContainer} in your {@link TileEntity} and forward the calls to an instance of this
+ * class for default multipart container logic.
```
No, TEs shouldnt implement any interface this needs to be a cap if anything.

asiekierka replied:
Do you expect EVERY SINGLE TE WHICH IMPLEMENTS THE CAP to also edit validate, invalidate and the other functions to match?

This is not an inventory. It's pretty important to have it hooked to most functions in a TE or in a Block.

amadornes replied:
Although I (again) agree with asie on this, turning it into a cap would remove the need for it to be implemented on the TE at all and would encourage people to use the default MultipartContainer instead. Capabilities will be much better :)

LexManos commented on src/main/java/net/minecraftforge/fmp/multipart/MultipartRedstoneHelper.java:
```diff
@@ -0,0 +1,152 @@
+        if (container == null) { return 0; }
+
+        IMultipart part = container.getPartInSlot(PartSlot.getFaceSlot(side));
+        if (part != null) { return part instanceof IRedstonePart ? ((IRedstonePart) part).getStrongSignal(side) : 0; }
```
instaneof bad bad bad

amadornes replied:
This will be in IMultipart now :)

LexManos commented on src/main/java/net/minecraftforge/fmp/util/IWorldLocation.java:
```diff
@@ -0,0 +1,16 @@
+public interface IWorldLocation
+{
+    
+    public World getWorldIn();
```
whats with the "In" parts?

asiekierka replied:
This was changed due to obfuscation bugs, but I'm not sure on the details anymore.

amadornes replied:
We'd need something like @RainWarrior's tree obfuscator if we want something like this to work without custom names (or at least that's what he told me). Right now it has "in" at the end to prevent the names from matching the ones in classes like TileEntity and breaking obfuscation.

LexManos commented on src/main/java/net/minecraftforge/fml/common/Loader.java:
```diff
@@ -941,20 +941,22 @@ boolean checkRemoteModList(Map<String, String> modList, Side side)
      * @return List with the mapping results.
      */
-    public List<String> fireMissingMappingEvent(Map<ResourceLocation, Integer> missingBlocks, Map<ResourceLocation, Integer> missingItems, boolean isLocalWorld, Map<ResourceLocation, Integer[]> remapBlocks, Map<ResourceLocation, Integer[]> remapItems)
+    public List<String> fireMissingMappingEvent(Map<ResourceLocation, Integer> missingBlocks, Map<ResourceLocation, Integer> missingItems, Map<ResourceLocation, Integer> missingMultiparts, boolean isLocalWorld, Map<ResourceLocation, Integer[]> remapBlocks, Map<ResourceLocation, Integer[]> remapItems, Map<ResourceLocation, Integer[]> remapMultiparts)
```
We really need to move this to a map or something so we can add everything we need without changing the signature.... @cpw ?

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/PersistentRegistryManager.java:
```diff
@@ -476,6 +496,7 @@ else if (action == FMLMissingMappingsEvent.Action.BLOCKONLY)
                 {
                     defaulted.add(remap.name);
+                    defaultedShort.add(remap.name);
                 }
```
...?

asiekierka replied:
defaultedShort gives users a count of the *names* which were removed - the default count doesn't take into account that, for instance, blocks and items with the same ID are usually the same thing. It's for user friendliness.

LexManos commented on src/main/java/net/minecraftforge/fmp/multipart/MultipartRegistry.java:
```diff
@@ -58,10 +59,10 @@ public static void registerPartFactory(IAdvancedPartFactory factory, ResourceLoc
         for (ResourceLocation part : parts)
         {
-            partProviders.put(part, factory);
+            GameRegistry.findRegistry(MultipartRegistryEntry.class).register(new MultipartRegistryEntry(factory).setRegistryName(part));
```
CACHE THTAT SHIT.

amadornes replied:
Will do... *runs away from really mad Lex*

LexManos commented on src/main/java/net/minecraftforge/fmp/multipart/MultipartRegistry.java:
```diff
@@ -123,65 +124,65 @@ public static ResourceLocation getPartType(IMultipart part)
     {
-        IAdvancedPartFactory factory = partProviders.get(partType);
+        IAdvancedPartFactory factory = GameRegistry.findRegistry(MultipartRegistryEntry.class).getValue(partType);
         return factory == null ? null : factory.createPart(partType, tag);
```
CACHE THAT SHIT

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/GameRegistry.java:
```diff
@@ -288,6 +298,46 @@ public static int getFuelValue(ItemStack itemStack)
+    /**
+     * Links a set of parts to an {@link IAdvancedPartFactory} that can produce them.
+     */
+    public static void registerPartFactory(IAdvancedPartFactory factory, ResourceLocation... names)
```
Why are you adding names? What are these functions doing? also why do you need a IPartFactory and an IAdvancedPartFactory?
We're trying to move away from register*** functions and just use the register(IForgeRegistryEntry)

asiekierka replied:
To create different parts based on NBT and/or packet data - see https://github.com/CharsetMC/Charset/blob/1.9/src/main/java/pl/asie/charset/wires/logic/PartWireProvider.java

amadornes replied:
Exactly what asie said. IPartFactory is there in case you don't need to handle both, to prevent code duplication or people routing one method to the other, which is a bit of a dirty hack since you'd need to cast null.

LexManos commented on patches/minecraft/net/minecraft/block/Block.java.patch:
```diff
@@ -1029,39 +1029,117 @@
++     * @return True if it can be rotated, false if it can't
 +     */
-+    public boolean rotateBlock(World world, BlockPos pos, EnumFacing axis)
++    public boolean canRotateBlockAround(World world, BlockPos pos, EnumFacing axis)
```
Why do we need a can function? we specificlaly removed that crap when we simplified the rotation.

asiekierka replied:
Wrenches would LOVE to return a FAIL when a rotation is not possible. Especially with the two-hand system now.

LexManos commented on patches/minecraft/net/minecraft/block/Block.java.patch:
```diff
@@ -1029,39 +1029,117 @@
++    
 +    /**
-+     * Get the rotations that can apply to the block at the specified coordinates. Null means no rotations are possible.
-+     * Note, this is up to the block to decide. It may not be accurate or representative.
```
None of these edits are nessasary... all you need is a ray trace on rotateBlock.... Much simpler...

amadornes replied:
While talking about it on the FC channel we came to the conclusion that the current system is too abstract for most cases where the player is not the one rotating the block. That's why they've been split into different methods. Also, getValidRotations() had a purpose at first but when rotateBlock() was generalized it lost all its meaning and most mods didn't even check it before performing a rotation.
Now we have a method that ensures the block will be rotated around the specified axis for external rotation systems, as well as one that can be a bit more abstract for things that want to handle rotation differently with a wrench (or do not want a wrench to rotate the block, but they want frames to do it, for example - which has been requested for a while).

LexManos commented on patches/minecraft/net/minecraft/block/Block.java.patch:
```diff
@@ -1116,6 +1194,21 @@
++     * @return If the recoloring was successful
++     */
++    public boolean recolorBlock(World world, BlockPos pos, EnumFacing side, net.minecraft.item.EnumDyeColor color, EntityPlayer player, RayTraceResult hit)
++    {
```
Player isnt needed, hit is.

asiekierka replied:
Player is useful, for instance when you want to restrict recoloring a block to certain players only.

amadornes replied:
Just like in the case of rotation, the old recolorBlock() is more generic and would affect the whole block, and the new one would affect only a specific part of it (if coded to do so). Also, it's nice to know the player that's trying to recolor it. I think it should be supplied whenever possible to avoid people trying to get it from the raytrace and doing messy things like that. As long as it's not completely unnecessary information (like https://xkcd.com/221/), it's always to know more information about an event. Who knows when it may come in handy!

LexManos commented on src/main/java/net/minecraftforge/fmp/ForgeMultipart.java:
```diff
@@ -18,11 +18,11 @@
+@Mod(modid = ForgeMultipart.MODID, name = ForgeMultipart.NAME, version = ForgeMultipart.VERSION)
 public class ForgeMultipart
 {
     
```
The entire @Mod shouldnt be needed anymore if it's going internal to forge then it's not a mod.

LexManos commented:
First pass of code is done basic gist is that a lot of stuff isn't needed.
This should probably be split into two PRs, one being the API the other being the implementation of the API.
Things need to be explained/documented better.
Capabilities need to be actually used instead of interfaces on hard classes everywhere.
Not to mention a lot of code style issues.
if (asdf)
  return blah;

not 
if (asdf){return blah}
also short circuiting methods instead of using inverted ifs is bad.

Overall this needs a LOT more cleanup/comments.

asiekierka commented:
>This should probably be split into two PRs, one being the API the other being the implementation of the API.

I don't think the two are entirely separable.

>Capabilities need to be actually used instead of interfaces on hard classes everywhere.

Some of these hard classes are required due to the assumptions that are made on, for instance, a multipart-containing block.

amadornes commented:
Alright! I'll to over it again today. Thanks for spending so much time looking for issues. The formatting ones were caused by the formatter (configured following cpw's recommendations) not be being fully configured and defaulting some settings. I'll have to check everything again, but it shouldn't be hard to fix :)
In terms of splitting API and implementation into different PRs, I'm not sure how feasible that'll be since they are very tightly bound. You'll have to tell me exactly what you want when I get home from school today :P

ProfessorProspector commented:
@amadornes in the mean time, would you mind putting a build of MCMP for 1.9.4 on curse? We're getting issues because our cables use multiparts and I don't believe there's a public build out. Not sure how long the issues with this PR are going to take, don't want to keep the people waiting too long.

LatvianModder commented:
@ProfessorProspector: That wont happen, because there would later be 2 multipart apis that do the same

amadornes commented:
@ProfessorProspector I didn't put out a build for that exact reason. I didn't want mods which depend on MCMP to release builds that use the 1.9.4 version to avoid clashing when it got added to Forge. The only reason it's on maven is so that people could update their mods to 1.9.4, but not for public use. It'll most likely be in by the time Lex decides to do a release build (so probably this weekend) :)

liach commented on src/main/java/net/minecraftforge/fmp/block/BlockCoverable.java:
```diff
@@ -0,0 +1,829 @@
+    }
+
+    @Override
+    public TileEntity createNewTileEntity(World worldIn, int meta)
```
To make this `final`?

asiekierka replied:
No. Definitely not.

liach replied:
Then why is there another `createNewTileEntityDefault` with same parameters below?
```java
    @SuppressWarnings("unchecked")
    public <T extends TileEntity & IMicroblockContainerTile> T createNewTileEntityDefault(World worldIn, int meta)
    {
        return (T) new TileCoverable();
    }
```

asiekierka replied:
Sounds like an artifact to me, can be removed?

amadornes replied:
This will all be gone in favour of capabilities

liach commented on src/main/java/net/minecraftforge/fmp/block/BlockMultipartContainer.java:
```diff
@@ -0,0 +1,561 @@
+ * {@link Multipart} or implement {@link IMultipart}. If you only need microblock support, look into
+ * {@link BlockCoverable}.
+ */
+public final class BlockMultipartContainer extends Block implements ITileEntityProvider
```
This needs to extend `BlockContainer` but it does not.

tterrag1098 replied:
Nothing needs to extend BlockContainer. Ever. Why would it?

liach replied:
The javadoc above says:
`A final class that extends {@link BlockContainer} and represents a block which can contain any kind of multipart.`

Lordmau5 replied:
Probably was a BlockContainer in earlier revisions and now it's just changed to a Block + ITileEntityProvider implementation :P

amadornes replied:
Yeah, outdated javadocs. Already fixed

liach commented on src/main/java/net/minecraftforge/fmp/util/MCMPEventHandler.java:
```diff
@@ -0,0 +1,56 @@
+import net.minecraftforge.fml.relauncher.SideOnly;
+import net.minecraftforge.fmp.multipart.MultipartRegistry;
+
+public class MCMPEventHandler
```
Should this be moved to forge internal handler?

amadornes replied:
There's no real need for that... It doesn't hurt anybody. It's been moved since the last commit, though. It's in net.minecraftforge.fmp and the util package is gone.

amadornes commented:
That was a long session of replying to Lex's "NOOOOOOO", "CACHE THAT SHIT." and "Absolutely none of these patches are necessary." comments :P
Time to fix what I haven't had time to look at yet! :D

liach commented:
@amadornes I also suggest you to encapsulate `net.minecraftforge.fmp.event.PartEvent`, let modders use accesors to get the `IMultiPart` instance, to be consistent with other forge events.

amadornes commented:
@liach Lex suggested changing that a few hours ago. It's already fixed :)

amadornes commented:
Almost ready for round two! I'll probably push the changes tomorrow :)

amadornes commented:
I promised I'd do it and I delivered. A bit later than I expected... but I delivered :P
Still need to split it into two different PRs because Lex doesn't want them to be in the same one. It's going to be a tough task if I want the API to compile without the implementation because they are very tightly linked, but if @RainWarrior says it's okay if they don't compile separately I may just leave them that way. @LexManos: What do you think? Mind if they don't compile separately?

liach commented on src/main/java/net/minecraftforge/fmp/ForgeMultipartModContainer.java:
```diff
@@ -0,0 +1,162 @@
+        meta.authorList = Arrays.asList("amadornes");
+        meta.description = "Forge Multipart provides the ability for mods to have more than one \"block\" " +
+                "in the same blockspace. ";
+        meta.url = "https://github.com/MinecraftForge/comingsoon/wiki";
```
Eh... This is ...

amadornes replied:
That is how Forge and FML are loaded. The wiki URL is temporary because we do not have one yet.

LexManos replied:
The API does not need it's own mod container.
The FML one is in for purely legacy reason and will *eventually* be going away.

liach commented on src/main/java/net/minecraftforge/fmp/ForgeMultipartModContainer.java:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    @Subscribe
+    public void postInit(FMLPostInitializationEvent event)
```
This method is not necessary.

amadornes replied:
True that, I'll remove it in round 2 :)

amadornes commented:
Since I haven't got a reply, I'll follow @RainWarrior's suggestion and split them even if they don't compile separately. Updated commits should be up in the next few days (hopefully tomorrow - if not, they'll have to wait a couple more days... blame finals).

bdew commented:
I'm not sure what the point of splitting the API and the implementation in this case would be, since they are meaningless without each other. Seems like busywork to me.

heldplayer commented:
@bdew Exactly, there's no reason they should be separated as they are each other.

modmuss50 commented:
There is no point in splitting them just to join the back up again.

asiekierka commented:
There's no point in splitting the API and the implementation when the API and implementation are closely tied together in many areas, such as with a multipart system. Where does one draw the line?

RainWarrior commented:
The point is to review the 10k-line PR in digestible parts; reasonable bounds for splitting were discussed, one of them is API + implementation.

asiekierka commented:
@RainWarrior Ah. So it's about reviewing it. Couldn't one, then, just make a list of which files belong to what?

RainWarrior commented:
@asiekierka Yes. Now, what's a good way to mark files from a commit as members of 2 sets, and show it conveniently on the github? :D

asiekierka commented:
@RainWarrior Obviously, it's using a variation of the Mechanical Turk.

amadornes commented:
Did I fuck up? I probably did. Opening another PR to master.
